source/AI.cpp

```diff
@@ -5161,3 +5179,93 @@ void AI::UpdateOrders(const Ship &ship)
 
 	it->second.Update(ship);
 }
+
+
+
+RoutePlan AI::GetRoutePlan(Ship &ship, const System *targetSystem)
+{
+	const System *from = ship.GetSystem();
+	const Government *gov = ship.GetGovernment();
+
+	// Look for an existing distance map for this combination of inputs before calculating a new one
+	// Make the key for the cache
+	const JumpType driveCapability = ship.JumpNavigation().HasJumpDrive() ? JumpType::JUMP_DRIVE : JumpType::HYPERDRIVE;
+
+	// A cached route that could be used for this ship could depend on the wormholes which this ship can
+	// travel through. Find the intersection of all known wormhole required attributes and the attributes
+	// which this ship satisfies.
+	vector<string> wormholeKeys;
+	const auto &shipAttributes = ship.Attributes();
+	for(auto requirement : universeWormholeRequirements)
+		if(shipAttributes.Get(requirement))
+			wormholeKeys.emplace_back(requirement);
+
+	// Search for a cached solution for this route.
+	auto key = RouteCacheKey(from, targetSystem, gov, ship.JumpNavigation().JumpRange(),
+		driveCapability, wormholeKeys);
+
+	RoutePlan route;
+	auto it = routeCache.find(key);
+	if(it == routeCache.end())
+	{
+		route = RoutePlan(ship, *targetSystem, ship.IsYours() ? &player : nullptr);
+		routeCache.emplace(key, route);
+	}
+	else
+	{
+		route = RoutePlan(it->second);
+	}
+
+	return route;
+}
+
+
+
+AI::RouteCacheKey::RouteCacheKey(
+	const System *from, const System *to, const Government *gov, double jumpDistance,
+	JumpType jumpType, const vector<string> &wormholeKeys)
+		: from(from), to(to), gov(gov), jumpDistance(jumpDistance),
+			jumpType(jumpType), wormholeKeys(wormholeKeys)
+{
+}
+
+
+
+size_t AI::RouteCacheKey::HashFunction::operator()(RouteCacheKey const &key) const
+{
+	// Used by unordered_map to determine equivalence.
+	int shift = 0;
+	size_t hash = std::hash<string>()(key.from->TrueName());
+	hash ^= std::hash<string>()(key.to->TrueName()) << ++shift;
+	hash ^= std::hash<string>()(key.gov->TrueName()) << ++shift;
+	hash ^= std::hash<int>()(key.jumpDistance) << ++shift;
+	hash ^= std::hash<int>()(static_cast<std::size_t>(key.jumpType)) << ++shift;
+	for(string k : key.wormholeKeys)
+	{
+		hash ^= std::hash<string>()(k) << ++shift;;
+	}
+	return hash;
+}
+
+
+
+bool AI::RouteCacheKey::operator==(const RouteCacheKey &other) const
+{
+	// Used by unordered_map to determine equivalence.
+	bool eq = true;
+	eq &= from == other.from;
+	eq &= to == other.to;
+	eq &= gov == other.gov;
+	eq &= jumpDistance == other.jumpDistance;
+	eq &= jumpType == other.jumpType;
+	eq &= wormholeKeys == other.wormholeKeys;
+	return eq;
```

Comment:
If we just stick this all in the return statement then it'll shortcut as soon as one is false.
```suggestion
	return from == other.from
		&& to == other.to
		&& gov == other.gov
		&& jumpDistance == other.jumpDistance
		&& jumpType == other.jumpType
		&& wormholeKeys == other.wormholeKeys;
```